squidpy/gr/_sepal.py

```diff
@@ -0,0 +1,341 @@
+        if entropy_arr[i] <= thres:
+            break
+
+    return np.nonzero(entropy_arr <= thres)[0]
+
+
+# taken from https://github.com/almaan/sepal/blob/master/sepal/models.py
+@njit(parallel=False, fastmath=True)
+def _laplacian_rect(
+    centers: np.ndarray,
+    nbrs: np.ndarray,
+    h: np.ndarray,
+) -> np.float_:
+    """Laplacian approx rectilinear grid."""
+    d2f = nbrs - 4 * centers
+    d2f = d2f / h ** 2
+
+    return d2f
+
+
+# taken from https://github.com/almaan/sepal/blob/master/sepal/models.py
+@njit(parallel=False, fastmath=True)
+def _laplacian_hex(
+    centers: np.ndarray,
+    nbrs: np.ndarray,
+    h: np.ndarray,
+) -> np.float_:
+    """Laplacian approx hexagonal grid."""
+    d2f = nbrs - 6 * centers
+    d2f = d2f / h ** 2
+    d2f = (d2f * 2) / 3
+
+    return d2f
+
+
+# taken from https://github.com/almaan/sepal/blob/master/sepal/models.py
+@njit(parallel=False, fastmath=True)
+def _entropy(
+    xx: np.ndarray,
+) -> np.float_:
+    """Entropy of array."""
+    xnz = xx[xx > 0]
+    xs = np.sum(xnz)
+    xn = xnz / xs
+    xl = np.log(xn)
+    return (-xl * xn).sum()
+
+
+def _compute_idxs(
+    g: spmatrix, spatial: np.ndarray, sat_thres: int, metric: str = "l1"
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """Get saturated and unsaturated nodes and nhood indexes."""
+    sat, unsat = _get_sat_unsat_idx(g.indptr, g.shape[0], sat_thres)
+
+    sat_idx, nearest_sat, un_unsat = _get_nhood_idx(sat, unsat, g.indptr, g.indices, sat_thres)
+
+    # compute dist btwn remaining unsat and all sat
+    dist = pairwise_distances(spatial[un_unsat], spatial[sat], metric=metric)
+    # assign closest sat to remaining nearest_sat
+    nearest_sat[np.isnan(nearest_sat)] = sat[np.argmin(dist, axis=1)]
+
+    return sat, sat_idx, unsat, nearest_sat.astype(np.int32)
+
+
+@njit(parallel=False)
+def _get_sat_unsat_idx(g_indptr: np.ndarray, g_shape: int, sat_thres: int) -> Tuple[np.ndarray, np.ndarray]:
+    """Get saturated and unsaturated nodes based on thres."""
+    n_indices = np.diff(g_indptr)
+    unsat = np.arange(g_shape)[n_indices < sat_thres]
+    sat = np.arange(g_shape)[n_indices == sat_thres]
+
+    return sat, unsat
+
+
+@njit(parallel=False)
+def _get_nhood_idx(
+    sat: np.ndarray, unsat: np.ndarray, g_indptr: np.ndarray, g_indices: np.ndarray, sat_thres: int
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Get saturated and unsaturated nhood indexes."""
+    # get saturated nhood indices
+    sat_idx = np.zeros((sat.shape[0], sat_thres))
+    for idx in np.arange(sat.shape[0]):
+        i = sat[idx]
+        s = slice(g_indptr[i], g_indptr[i + 1])
+        sat_idx[idx] = g_indices[s]
+
+    # get closest saturated of unsaturated
+    nearest_sat = np.zeros(unsat.shape) * np.nan
+    for idx in np.arange(unsat.shape[0]):
+        i = unsat[idx]
+        s = slice(g_indptr[i], g_indptr[i + 1])
+        unsat_neigh = g_indices[s]
```

Comment:
I'd remove the slice and just use `g_indices[g_indptr[i]: g_indptr[i + 1]]` unless numba complains, same as above.